zqd/packet/packet.go

```diff
@@ -0,0 +1,236 @@
+			return p.err
+		case event := <-w.Events:
+			if event.Op == fsnotify.Create && filepath.Ext(event.Name) == ".log" {
+				return nil
+			}
+		case err := <-w.Errors:
+			return err
+		}
+	}
+}
+
+func (p *IngestProcess) writeIndexFile(idx *pcap.Index) error {
+	idxpath := p.space.DataPath(IndexFile)
+	tmppath := idxpath + ".tmp"
+	f, err := os.Create(tmppath)
+	if err != nil {
+		return err
+	}
+	if err := json.NewEncoder(f).Encode(idx); err != nil {
+		f.Close()
+		return err
+	}
+	f.Close()
+	return os.Rename(tmppath, idxpath)
+}
+
+func (p *IngestProcess) slurp(ctx context.Context) (*pcap.Index, error) {
+	pcapfile, err := os.Open(p.pcapPath)
+	if err != nil {
+		return nil, err
+	}
+	defer pcapfile.Close()
+	zeekproc, zeekwriter, err := p.startZeek(ctx, p.logdir)
+	if err != nil {
+		return nil, err
+	}
+	indexwriter := pcap.NewIndexWriter(10000)
+	w := io.MultiWriter(zeekwriter, indexwriter, p)
+	if _, err := io.Copy(w, pcapfile); err != nil {
+		return nil, err
+	}
+	// Now that all data has been copied over, close stdin for zeek process so
+	// process gracefully exits.
+	if err := zeekwriter.Close(); err != nil {
+		return nil, err
+	}
+	if err := zeekproc.Wait(); err != nil {
+		return nil, err
+	}
+	index, err := indexwriter.Close()
+	if err != nil {
+		return nil, err
+	}
+	return index, nil
+}
+
+// PcapReadSize returns the total size in bytes of data read from the underlying
+// pcap file.
+func (p *IngestProcess) PcapReadSize() int64 {
+	return atomic.LoadInt64(&p.pcapReadSize)
+}
+
+// Err returns the an error if an error occurred while the ingest process was
+// running. If the process is still running Err will wait for the process to
+// complete before returning.
+func (p *IngestProcess) Err() error {
+	<-p.done
+	return p.err
+}
+
+// Done returns a chan that emits when the ingest process is complete.
+func (p *IngestProcess) Done() <-chan struct{} {
+	return p.done
+}
+
+func (p *IngestProcess) writeData(ctx context.Context) error {
+	files, err := filepath.Glob(filepath.Join(p.logdir, "*.log"))
+	// Per filepath.Glob documentation the only possible error would be due to
+	// an invalid glob pattern. Ok to panic.
+	if err != nil {
+		panic(err)
+	}
+	// convert logs into sorted bzng
+	zr, err := scanner.OpenFiles(resolver.NewContext(), files...)
+	if err != nil {
+		return err
+	}
+	defer zr.Close()
+	// For the time being, this endpoint will overwrite any underlying data.
+	// In order to get rid errors on any concurrent searches on this space,
+	// write bzng to a temp file and rename on successful conversion.
+	bzngfile, err := p.space.CreateFile("all.bzng.tmp")
+	if err != nil {
+		return err
+	}
+	zw := bzngio.NewWriter(bzngfile)
+	const program = "sort -limit 10000000 ts"
+	if err := search.Copy(ctx, zw, zr, program); err != nil {
+		// If an error occurs here close and remove tmp bzngfile, lest we start
+		// leaking files and file descriptors.
+		bzngfile.Close()
+		os.Remove(bzngfile.Name())
+		return nil
```

Comment:
@mattnibs shouldn't this return the error?

Reply:
Whoops, yeah. There was one other error I made in this pr. I'll file a follow up error that fixes that and this.